(PCP-645) implement v2 protocol

wkalt opened the pull request:
This has been tested manually against pxp-agent and clj-pxp-puppet, as well as a basic text-based Websocket client. I also ran it in pxp-agent acceptance tests with modest modification to test setup (https://github.com/puppetlabs/pxp-agent/pull/529).

It'll need a squash before merge. It also depends on https://github.com/puppetlabs/clj-pcp-common/pull/32, which needs to be merged and released for unit tests to pass.

wkalt commented on src/puppetlabs/pcp/broker/connection.clj:
```diff
@@ -43,28 +32,57 @@
+   :remoteaddress (ws->remote-address ws)})
+
+;; TODO this doesn't actually work for supplied client types. figure out what
+;; to do.
```
need to fix

wkalt commented on test/utils/puppetlabs/pcp/testutils/client.clj:
```diff
@@ -25,7 +27,7 @@
   (send! [_ message]
     (http/send ws-client :byte (message/encode message)))
-  (recv! [this] (recv! this (* 10 5 1000)))
+  (recv! [this] (recv! this (* 3 1000)))
```
this seemed to make the tests run faster, even when they all passed

puppetcla commented:
CLA signed by all contributors.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -139,17 +162,20 @@
+        {:keys [websocket]} connection
+        version (connection/ws->protocol-version websocket)
+        encoded-message (encode error-msg version)]
+    (websockets-client/send! websocket encoded-message)
```
Links to my comment at https://github.com/puppetlabs/clj-pcp-common/pull/31#discussion_r92051978. I guess I'm ok with this division if it works, just seems unclear to me whether it should work.

MikaelSmith replied:
`send!` maps to `sendBytes` or `sendText` at https://github.com/puppetlabs/trapperkeeper-webserver-jetty9/blob/master/src/puppetlabs/trapperkeeper/services/webserver/experimental/jetty9_websockets.clj#L25-L36. I'd really rather send text as text, and have onText expect text (rather than a byte array).

It's ok if onText only works with the v2 endpoint, and onBytes only works with the v1 endpoint.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -165,27 +191,29 @@
-    (if-let [first-target (first targets)]
-      (or (not= 1 (count targets)) (p/uri-wildcard? first-target)))))
+  (if-let [target (:target message)]
+    (if-let [first-target (first target)]
```
Seems like it might not work right if `:target` is a single URI (rather than an array).

wkalt replied:
yeah, the array was a brain fart. Will go back through and convert.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -237,97 +245,95 @@
+  (let [ws (:websocket connection)
+        version (connection/ws->protocol-version ws)
+        id (:id request)
+        requester-uri (:sender request)
```
We shouldn't just accept the sender URI, we should generate it from the websocket connection. Your current code allows masquerading as another client. Huh, I was sure this was checked, but now I can't find where.

MikaelSmith replied:
Ah right, it's `authenticated?`. Nevermind, this code should be safe.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -237,97 +245,95 @@
 (s/defn make-inventory_response-data-content :- p/InventoryResponse
-  [uris]
-  {:uris uris})
+  [uris broker]
```
We should pass in version, rather than broker. We should also make sure querying uris and version are done as one atomic query.

wkalt commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -139,17 +162,20 @@
-        encode (:encode codec)]
-    (websockets-client/send! websocket (encode error-msg))
+        error-msg (-> (message/make-versioned-message
+                        "v2.0"
```
right. everything internal is presumed to be on the latest version. It may be worth sticking latest version in a var for maintainability though.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -497,12 +498,13 @@
         ;; TODO(richardc): this could use a different message_type to
         ;; indicate an encoding error rather than a processing error
-        (send-error-message nil (i18n/trs "Could not decode message") connection)))))
+        (send-error-message nil (i18n/trs "Could not decode message") {:websocket ws}))))
```
This looks wrong. Isn't `ws` already the websocket?

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -514,8 +516,7 @@
       (when-let [connection (process-message! broker bytes ws)]
-        (assert (instance? Connection connection))
-        (.put (:connections broker) ws connection)))))
+        (assert (instance? Connection connection))))))
```
We could probably drop the `when-let` and `assert`.

wkalt replied:
I left this in place because some tests depend on it today. We can probably work it out though.

MikaelSmith commented on src/puppetlabs/pcp/broker/connection.clj:
```diff
@@ -43,28 +32,57 @@
+    :v2
+    (let [path (.. ws getSession getUpgradeRequest getRequestURI getPath)]
+      (case path
+        "/pcp/v1.0" :v1
```
I'd rather not hard-code these. They definitely won't match this in PE. Also, drop vNext, it's no longer used.

They should query the webroutes service for the relevant path.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -82,33 +110,28 @@
+  [broker :- Broker
+   ws :- Websocket]
+  (let [{:keys [common-name] :as connection} (connection/make-connection ws)
+        path (str "pcp://" common-name "/agent")]
```
You have `ws->uri`, we should use that. Also, `agent` is just the default for v1 messages. v2 should extract it from the path (using similar logic to remove the prefix as I noted for `ws->protocol-version`).

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -82,33 +110,28 @@
-  (.remove (:connections broker) ws))
+  [broker :- Broker
+   common-name]
+  (when-let [path (str "pcp://" common-name "/agent")]
```
Same note about using `ws->uri` and `agent` only applying to v1 clients.

wkalt commented on test/utils/puppetlabs/pcp/testutils/client.clj:
```diff
@@ -39,22 +41,22 @@
-  (let [uri                 (or uri (str "pcp://" certname "/test"))
-        association-request (modify-association (make-association-request uri))
+           version "v2.0"}}]
+  (let [uri                 (or uri (str "pcp://" certname "/agent"))
```
on revisiting, I think this may be acceptable for test utils

wkalt commented:
@MikaelSmith I think I've addressed all your comments. There are updates on clj-pcp-common, and a new PR on trapperkeeper-webserver-jetty9 here: https://github.com/puppetlabs/trapperkeeper-webserver-jetty9/pull/156

I've left the commits split out here in case it's useful to you; I'm assuming they'll be squashed prior to merge. Sorry about anything I've missed and have an excellent holiday!

MikaelSmith commented:
I have some work on this I'll put up tomorrow. I need to fix a few tests and clean it up.

MikaelSmith commented:
Still a work in progress. I added a commit based on changes proposed in https://github.com/puppetlabs/clj-pcp-common/pull/32. Also tested with various clients and made fixes. Need to rework tests now, and I'll want to run this through the full pxp-agent test suite before merging as well.

It also doesn't quite match the pcp-specifications definition of v2 yet.

puppetcla commented:
CLA signed by all contributors.

MikaelSmith commented:
I plan to update the docs folder still.

MikaelSmith commented:
@puppetlabs/pcp-broker-maintainers I think this is ready for review. I've tested it manually against pxp-agent and clj-pxp-puppet, as well as a basic text-based Websocket client. I also ran it in pxp-agent acceptance tests with modest modification to test setup (https://github.com/puppetlabs/pxp-agent/pull/529).

It'll need a squash before merge. It also depends on https://github.com/puppetlabs/clj-pcp-common/pull/32, which needs to be merged and released for unit tests to pass.

wkalt commented on src/puppetlabs/pcp/broker/connection.clj:
```diff
@@ -21,6 +22,7 @@
 (s/defrecord Connection
              [websocket :- Websocket
               remote-address :- s/Str
+              codec :- Codec
```
I don't understand the reason for tracking this in every connection when the codecs used in practice are so predictable.

wkalt commented on test/integration/puppetlabs/pcp/broker/service_test.clj:
```diff
@@ -263,33 +271,35 @@
                      (is (= "http://puppetlabs.com/inventory_response" (:message_type response)))
-                     (is (= (:id request) (:in_reply_to response)))
+                     (if (= version "v1.0")
+                       (is (= (:id request) (:in-reply-to response)))
```
we should update the v1 spec to mention in-reply-to

wkalt commented on test/unit/puppetlabs/pcp/broker/message_test.clj:
```diff
@@ -0,0 +1,40 @@
+                    :target "pcp://gangoffour/entity"})]
+      (is (= (-> message
+                 v2-encode
+                 v2-decode))))))
```
this test and the one above it will always pass

wkalt commented on test/integration/puppetlabs/pcp/broker/service_test.clj:
```diff
@@ -121,21 +121,28 @@
-  [message version]
-  "Assert that the message is a successful association response."
+  (let [data (client/get-data message version)]
+    (if (= "v1.0" version)
```
like you mentioned in the other patch it would be preferable to switch to a keyword representation of version throughout -- not sure if we want to take that step at this point.

MikaelSmith replied:
For testing, this is fine.

wkalt commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -433,27 +400,37 @@
+   uri :- p/Uri]
+  (cond-> message
+    (not (:sender message)) (assoc :sender uri)
+    (not (:target message)) (assoc :target "pcp:///server")))
```
this could be simplified with the new assoc-when function in utils

wkalt commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -346,17 +271,19 @@
-         :data response-data})
-       connection)))
+    (let [requester-uri (:sender message)
+          response-data (make-inventory_response-data-content broker (:query data))
```
pretty minor but these underscores bugged me a bit before -- is anyone attached to them?

MikaelSmith replied:
They correspond to an actual keyword (`inventory_response`) and I'd rather avoid too many changes in a PR this complex.

wkalt commented on src/puppetlabs/pcp/broker/message.clj:
```diff
@@ -0,0 +1,91 @@
+                     (dissoc :expires :destination_report)
+                     validate-not-multicast
+                     (rename-keys {:in-reply-to :in_reply_to})
+                     (rename-keys {:targets :target})
```
we can consolidate this line and the one above

wkalt commented on src/puppetlabs/pcp/broker/message.clj:
```diff
@@ -0,0 +1,91 @@
+                     (rename-keys {:in-reply-to :in_reply_to})
+                     (rename-keys {:targets :target})
+                     (update-when [:target] first)
+                     (dissoc :_chunks))
```
this can be consolidated with line 34

wkalt commented on src/puppetlabs/pcp/broker/message.clj:
```diff
@@ -0,0 +1,91 @@
+(s/defn v1->v2 :- m2/Message
+  "Transform a v1 Message to a v2 Message."
+  [v1-msg :- m1/Message]
+  (let [data' (m1/get-json-data v1-msg)
```
may as well just call this "data"

wkalt replied:
ah, never mind -- I see what you're doing below. Conventionally I'd expect data' to be bound after data.

wkalt replied:
(likewise with message)

wkalt replied:
see https://en.wikipedia.org/wiki/Prime_(symbol)#Use_in_mathematics.2C_statistics.2C_and_science for background

wkalt commented on src/puppetlabs/pcp/broker/message.clj:
```diff
@@ -0,0 +1,91 @@
+
+(s/defn v2->v1 :- m1/Message
+  "Transform a v2 Message to a v1 Message."
+  [{:keys [message_type data] :as v2-msg :- m2/Message}]
```
this should be
```
[{:keys [message_type data] :as v2-msg} :- m2/Message]
```
for the validation to work

wkalt replied:
may as well destructure in_reply_to here too to use it below

wkalt commented on test/integration/puppetlabs/pcp/broker/service_test.clj:
```diff
@@ -237,43 +232,37 @@
+                                                    :version version)]
+                   (let [pcp-response (client/recv! client)
+                         close-websocket-msg (client/recv! client)]
+                     (is (is-association_response pcp-response version false "Session already associated"))
```
the is here will cause redundant failures

wkalt commented on test/integration/puppetlabs/pcp/broker/service_test.clj:
```diff
@@ -442,99 +433,85 @@
+                       :targets ["pcp://client02.example.com/agent"]
+                       :destination_report true
+                       :message_type "greeting"})
+            message (m1/set-json-data message' "Hello")]
```
also seems like the double binding isn't needed here

wkalt commented on test/utils/puppetlabs/pcp/testutils/client.clj:
```diff
@@ -24,8 +28,11 @@
   (send! [_ message]
-    (http/send ws-client :byte (message/encode message)))
-  (recv! [this] (recv! this (* 10 5 1000)))
+    (let [version (if (:_chunks message) :v1 :v2)] ;; TODO do something better here
```
this TODO may or may not apply -- if presence of :_chunks is sufficient we can remove it.

wkalt commented:
took a pass through. mostly minor things, and +1 on the renamespacing stuff. Reintroduction of codecs feels like a smell to me, but I never understood why they were there so I may be missing some background.

MikaelSmith commented:
@wkalt the codec is a mechanism to associate connection with a version, so we don't have to pass it through in all functions that need to know, and removes conditional logic (call a function and pass in the version to get the correct callback). It more directly applies the thing we care about - conversion to a common format - rather than a version that can be associated with a conversion function. If we're worried about memory (8 bytes per connection) I could switch to threading version through all the operations.

MikaelSmith commented:
Updated.

wkalt commented:
@MikaelSmith No, I'm not worried about the memory cost. I understand your perspective. The thing I like about version dispatch in encode/decode is that since version is already available in connections via the ws adapter, using it allows the whole codec concept to be dropped with a minor change. It's not a big deal though; this patch seems fine.

MikaelSmith commented:
@wkalt I don't know that it is. At least, the way we stash path for `ws->uri` doesn't provide the full path, just everything after the route. Possibly the way we were retrieving it before (`getSession getUpgradeRequest getPath`) would allow comparing to the route.

MikaelSmith commented on test/unit/puppetlabs/pcp/broker/core_test.clj:
```diff
@@ -210,138 +152,101 @@
             (is (not (realized? @closed)))
-            (is (= :associated (:state connection)))
-            (is (= "pcp://localhost/controller" (:uri connection)))))
-        (testing "Associates a client already associated on a different session, but disconnects the first"
```
Actually, rejecting the new one is tricky to do, and I don't see a reason to change from the old behavior. So going with superseding the old one.

mruzicka commented on src/puppetlabs/pcp/broker/websocket.clj:
```diff
@@ -0,0 +1,48 @@
+  [ws]
+  (try+
+   (second
+    (str/split (.. (websockets-client/remote-addr ws) (toString)) #"/"))
```
We should be able to extract the address in a more straightforward way than parsing it from a string using a regexp.

mruzicka replied:
We should be able to extract the address in a more straightforward way than parsing it from a string using a regexp.

MikaelSmith replied:
Yeah. I think this was already there (just moved to a new function) so I didn't mess with it. There are ways to construct it out of hostname and port; I looked at that when I was adding to trapperkeeper-webserver-jetty9, but decided not to change it because it'd break existing behavior.

mruzicka commented:
In the end only one minor suggestion survived my long review, and it's a minor one. So let's move forward with this and address it later.

MikaelSmith commented:
Sounds good, thanks.
